src/Lecture2.hs

```diff
@@ -268,7 +327,15 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval _ (Lit n) = Right n
+eval vs (Var varName) = case lookup varName vs of
+  Just n -> Right n 
+  Nothing -> Left $ VariableNotFound varName
+
+eval vs (Add e1 e2) = case (eval vs e1, eval vs e2) of
+                        (Left err , _) -> Left err
+                        (_ , Left err) -> Left err
+                        (Right r1, Right r2) -> Right $ r1 + r2
```

Comment:
That's a great case matching! 👏🏻 
Once you learn about Monads, you'll be able to reduce this function to a single line ✂️